### DISCUSSION TITLE
LL-HLS solution for radio broadcast

### DISCUSSION THREAD

**Discussion by moleculezz** [bluenviron/mediamtx#1765] [Questions]
Hello,

I've been looking for a solution for a radio broadcast for a while now. What I want to accomplish is no disconnects when clients switch WiFi/Mobile networks and at most 6 sec latency. From what I understand LL-HLS would solve these issues, but up to now I wasn't able to find a project that solves this. This project seems like it might solve the problems I'm trying to overcome. 

The questions I had were the following:
1. What is the best protocol to generate the stream?
2. Can mediamtx function as both server and client for a stream?

I'm hoping to figure out an end to end solution from source of the stream to listeners.

**Comment by zapf2000** [comment]
Hi, I'm webradio broadcaster for almost 3 decades now and there are several solutions. If you only wanna broadcast audio, icecast is always an option. If you wanna go near realtime, you can choose mediamtx as server together with a LL-HLS webplayer (javascript), for example theoplayer. You can also have a look at ovenmediaengine/ovenplayer as a free open source solution for WebRTC/LL-HLS. It always depends on the demands you have (lossless quality streaming with FLAC or lossy with AAC/MP3/OPUS/OGG) and realtime or not.

**Comment by aler9** [comment]
In my opinion, LL-HLS is a better fit for for broadcasting with respect to WebRTC since the network load can be distributed with a HTTP CDN like CloudFront or Cloudflare, while there are few (or none?) CDNs for WebRTC. This allows to scale infinitely.

Regarding the media server, MediaMTX doesn't transcode, it takes what it receives and distributes it without any change. This has the advantage of using a negligible amount of computational power. If you instead prefer to re-encode the stream, there are other solutions.

**Reply by zapf2000** [reply]
a good media server doesn't need to transcode in any case, it just pass a single stream (even it has a lot of layers) to the public. But there are 2 criterias: do it instantly (no delay) ... do it reliable (no packet loss). so we need good methods for that... some kind of FEC (forward error correction), packet loss compensation (like done in haivision SRT) and a scaleable format (like done in AV1 or H265). Don't let the media server do all the work, do it yourself, but the media server has to support it.... and for sure needs to be able to cluster (origin/edges)

**Reply by aler9** [reply]
Transmission reliability is a protocol feature and is implemented in all protocols supported by MediaMTX:

* RTMP is TCP-based, therefore there are ackowledges and retries
* RTSP has packet reordering
* WebRTC has packet reordering, negative acknowledges and retransmission
* HLS is HTTP-based, therefore clients can retry requests indefinitely

SRT is a good protocol that will be added to the server in the near future, but it has no major advantages over WebRTC in terms of reliability (on the server side).

The server can achieve both no delay and no packet loss since it is not a player, it doesn't need to respect timing since it doesn't the decode stream, it just routes packets to players. Therefore, as soon as a packet is received correctly, it is sent out. If there are missing packets, packets are put in a queue until gaps are filled or the queue is full.

It's the client / player that has to perform the choice that you mention (no delay vs no packet loss), and this is usually achieved by setting the buffer size.

**Comment by moleculezz** [comment]
What do you recommend for the source of the stream? So from the radio station to mediamtx for example?

**Reply by aler9** [reply]
> What do you recommend for the source of the stream? So from the radio station to mediamtx for example?

In my opinion WebRTC, since it can retransmit lost packet, but only if you plan to use the Opus codec.
If you plan to use the AAC codec, you have to switch to RTSP, since it is not supported by WebRTC.

**Reply by moleculezz** [reply]
I plan to use AAC codec. What are some tools that can do RTSP?
I have the audio coming in from a Console through a Focusrite Scarlett 4i4 into the PC.

**Reply by zapf2000** [reply]
> I plan to use AAC codec. What are some tools that can do RTSP? I have the audio coming in from a Console through a Focusrite Scarlett 4i4 into the PC.

Do you like to broadcast audio only?

**Reply by moleculezz** [reply]
Mostly, but sometimes we do video as well to stream to Youtube or Facebook. 
EDIT: But this is on a different PC that handles the video.

**Reply by zapf2000** [reply]
> Mostly, but sometimes we do video as well to stream to Youtube or Facebook. EDIT: But this is on a different PC that handles the video.

Currently we use Ovenmediaengine for video and audio (MPEG-4 AVC ABR 1080p, 720p, 360p and AAC audio, also transcoded live to Opus for use in WebRTC), because we can ingest SRT via FFMPEG, gstreamer or OBS and also push RTMP to youtube or facebook.

In parallel, for audio we use the good old icecast2 servers (FLAC, MP3, AAC+V2 formats). Configured well, they can also achieve latencies of about 4-6 seconds and can be played on every custom webradio hardware (that was the major issue. It must be playable out of the box on custom hardware). The other issue was, that we didn't find any audio streaming server, that can handle FLAC format (24Bit/96KHz Hi-Res lossless audio streaming).

Relatively new we currently test mediamtx for use with AV1 video on webrtc, but my other post shows that the audio quality for ingesting webrtc is not configurable, which results in pretty good video, but very poor audio quality. Some improvements need to be done on the webrtc ingester. Currently mediamtx is the only open source server i know, which provides AV1 streaming in webrtc.
If you like to use mediamtx, you can ingest RTSP via FFMPEG, gstreamer or vlc. OBS and various other applications only seem to work properly with RTMP protocol, even they use FFMPEG libs. On listener (client) side you can use browser with video-tag or javascript players for ll-hls. I also tested RTMP streaming on a enigma2 satellite receiver without problems. Webrtc is also working fine, but you have to transcode aac to opus (can be done with FFMPEG on mediamtx server by calling ffmpeg in config file). What mediamtx cannot do directly, is transcode H264 for example (ABR video streaming).

**Reply by moleculezz** [reply]
I was considering using liquidsoap (which has ffmpeg support) or just plain FFMPEG. But liquidsoap has some nice fallback features for if the streaming breaks down.
EDIT: The 2 things I need are streaming and recording to have a backup of the stream.

**Reply by zapf2000** [reply]
> I was considering using liquidsoap (which has ffmpeg support) or just plain FFMPEG. But liquidsoap has some nice fallback features for if the streaming breaks down. EDIT: The 2 things I need are streaming and recording to have a backup of the stream.

Well then you should really have a look at icecast2 and butt as encoder (https://danielnoethen.de/butt/).
Encoding with FFMPEG is very nice too, but butt can also update metadata dynamically, which might be an issue too. Icecast also supports fallback mounts and clustering. Most webradios use it as primary streaming server.

**Reply by moleculezz** [reply]
We currently use icecast and butt on Windows, but I'm testing out a Linux server and butt has some compatibility issues with Jack2.

**Reply by zapf2000** [reply]
> We currently use icecast and butt on Windows, but I'm testing out a Linux server and butt has some compatibility issues with Jack2.

You can use butt with pulseaudio and create a virtual pulseaudio device for jack2. Just configure pulseaudio to use maximum capability of sample rate.
For creating a virtual device, you can use this command: pactl load-module module-jack-source client_name=playout connect=no
Then you can connect butt to the "playout" device.
![Screenshot_20230608_095020](https://github.com/bluenviron/mediamtx/assets/52130927/36418ee6-6d66-472c-9d51-08a9bcecf045)

[a comment by an automated account is not shown]
